### PR TITLE
DM: auxiliary surface gradient prediction for smoothness regularization

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Multi-task learning with auxiliary spatial gradient prediction (∂p/∂x, ∂p/∂y, ∂p/∂z) should regularize the encoder into learning smooth, physically consistent pressure fields. By explicitly supervising the gradient of the predicted surface pressure, we force the model to capture the underlying physics more faithfully — rather than interpolating point values independently.

This is grounded in physics-informed ML (Raissi et al., 2019, *Physics-informed neural networks*) adapted for graph networks: gradient consistency is a free regularization signal that comes from the mesh topology itself, requiring no additional data collection.

## Instructions

**Implementation — 4 steps:**

**Step 1: Compute ground-truth spatial gradients**

For each surface point, compute ground-truth gradient as weighted average over graph neighbours:
```python
# For each surface point i with neighbours j:
grad_x[i] = sum_j( (p[j] - p[i]) * (x[j] - x[i]) / dist(i,j)^2 ) / sum_j(1/dist(i,j))
grad_y[i] = sum_j( (p[j] - p[i]) * (y[j] - y[i]) / dist(i,j)^2 ) / sum_j(1/dist(i,j))
grad_z[i] = sum_j( (p[j] - p[i]) * (z[j] - z[i]) / dist(i,j)^2 ) / sum_j(1/dist(i,j))
```
Apply to all output channels (pressure, etc.) separately. Compute once per batch from ground-truth labels.

**Step 2: Add auxiliary gradient prediction head**

Add a 2-layer MLP (256d hidden, ReLU, output = 3 × num_output_channels) that takes the encoder node features and predicts the spatial gradient vector. This head branches off the same node embeddings used for the primary output head.

**Step 3: Compute auxiliary gradient loss**

```python
aux_loss = F.mse_loss(predicted_gradients, ground_truth_gradients)
```

Log this separately in W&B as `train/aux_gradient_loss` and `val/aux_gradient_loss`.

**Step 4: Combine losses**

```python
total_loss = primary_loss + aux_weight * aux_loss
```

where `aux_weight` is a hyperparameter.

---

**Trial 1** — `aux_weight=0.1` (moderate regularization):

```bash
cd target/icml2026
SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --grad-clip 0.5 --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --ema-decay 0.9995 \
  --aux-gradient-weight 0.1 \
  --wandb_group dm-auxiliary-gradient
```

**Trial 2** — `aux_weight=0.01` (lighter regularization):

Same command with `--aux-gradient-weight 0.01`.

**Smoke test first** — run 2 epochs to confirm the aux loss is non-NaN and logging correctly, then launch full runs for both trials.

**Important:** Log aux gradient loss separately in W&B (`train/aux_gradient_loss`, `val/aux_gradient_loss`). The primary metric to beat is still `val_primary/surface_rel_l2_pct`.

## Baseline

Current best metrics (DrivAerML):
- `val_primary/surface_rel_l2_pct`: **3.833%** (val) at epoch 511 — test **4.685%**
- Best PR: #3072 (eren — EMA=0.9995 + gc=0.5, 4L/512d/8H, AdamW lr=5e-4, T_max=30)
- Baseline W&B run: `ncl1dh88` (eren/dm-ema-9995-gc05)
- External target: <3.71% (AB-UPT) — gap: 0.123 pp
- Reproduce command:
  ```bash
  cd target/icml2026
  SENPAI_MAX_EPOCHS=9999 python train.py \
    --dataset drivaerml \
    --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
    --grad-clip 0.5 --enable-fourier \
    --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
    --epochs 999 \
    --batch-size 1 \
    --drivaerml-train-surface-points 50000 \
    --drivaerml-eval-surface-points 50000 \
    --max-train-batches 394 --max-eval-batches 200 \
    --ema-decay 0.9995
  ```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT kakashi:

## Results

### Trial 1: `aux_gradient_weight=0.1`

| Metric | Value |
|--------|-------|
| `val_primary/surface_rel_l2_pct` | 23.5% (best, ep20) |
| Outcome | **Diverged at epoch 22** |
| W&B run | `h611t6dg` |
| Peak memory | ~6 GB |

**Command:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=2 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --grad-clip 0.5 \
  --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --ema-decay 0.9995 --aux-gradient-weight 0.1 \
  --agent kakashi --wandb-name "kakashi/dm-aux-grad-w01" --wandb-group dm-auxiliary-gradient
```

**What happened:** Training diverged rapidly at epoch 20-22. Gradient norms exploded: 2.25 → 4.89 → 15.3 → 18.1 in successive epochs, overwhelming the gc=0.5 clipping. The aux gradient loss magnitude (~0.04) multiplied by weight 0.1 created a signal (~0.004) that was too strong relative to the primary loss, destabilizing the optimizer. Killed at epoch 22 to free GPU.

---

### Trial 2: `aux_gradient_weight=0.01`

| Metric | Baseline | Trial 2 (best) | Trial 2 (terminal) | Delta vs baseline |
|--------|----------|----------------|---------------------|-------------------|
| `val_primary/surface_rel_l2_pct` | **3.833%** (ep511) | **5.485%** (ep191) | 7.879% (ep196) | +1.652 pp worse |
| `val_surface/aux_gradient_loss` | N/A | 0.003720 (ep191) | 0.003776 (ep196) | — |
| `train/aux_gradient_loss` | N/A | 0.004370 (ep191) | 0.004425 (ep196) | — |
| W&B run | `ncl1dh88` | `exbz0pb4` | — | — |
| Peak memory | ~6 GB | ~6.5 GB | — | — |

**Command:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=3 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --grad-clip 0.5 \
  --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --ema-decay 0.9995 --aux-gradient-weight 0.01 \
  --agent kakashi --wandb-name "kakashi/dm-aux-grad-w001" --wandb-group dm-auxiliary-gradient
```

**What happened:** Training was stable for ~190 epochs with smooth convergence. Trajectory:

| Epoch | Best val % | Convergence rate |
|-------|-----------|------------------|
| 90 | 8.097% | — |
| 120 | 6.699% | ~0.047 pp/ep |
| 150 | 6.003% | ~0.023 pp/ep |
| 180 | 5.644% | ~0.012 pp/ep |
| 191 | 5.485% | ~0.014 pp/ep |
| 192+ | **Divergence** | Grad norms 0.26 → 1.09 |

At epoch 192, a sudden instability event occurred: grad_norm_mean spiked from 0.26 to 0.84 with 185 grad_clip_events (vs 10 at ep191), and val metric jumped from 5.485% to 5.580%. By ep195, grad_norm reached 1.09 with 223 clip events, and val had degraded to 7.311%. The EMA model could not absorb the damage — by ep196 val was 7.879% and still worsening. Killed at ep196.

**Analysis:**

1. **Negative result.** The auxiliary gradient prediction hypothesis did not improve performance. Even before divergence, the best val (5.485% at ep191) was 1.65pp above the baseline (3.833% at ep511). Extrapolating the pre-divergence convergence rate (0.012pp/ep), the run would have needed ~137 more epochs to even match baseline — and the convergence was decelerating.

2. **Delayed instability pattern.** Both trials showed the same failure mode at different timescales: w=0.1 diverged at ep22, w=0.01 diverged at ep192. The aux gradient signal appears benign during early training when the primary loss is large but becomes destabilizing once the model is well-trained and the loss landscape flattens. The kNN-based gradient targets may be noisy enough to inject harmful variance in the late-training regime.

3. **The aux gradient loss itself converged.** It dropped from ~0.005 to ~0.0037 over 190 epochs (train and val tracking closely), suggesting the MLP head was learning but the learned gradients weren't helping the primary task. The gradient regularization may have been constraining the model to smoother solutions that were suboptimal for the actual pressure distribution.

4. **Computational overhead.** The kNN gradient computation (chunked over 50K points with k=16) added ~10% overhead per step. Memory increase was modest (~0.5 GB).

## Suggested follow-ups

- **Gradient-free smoothness:** If the goal is smoother predictions, a Laplacian smoothness penalty (L2 on finite differences between neighboring predictions) might be more stable than explicit gradient supervision, since it doesn't require computing noisy kNN gradient targets.
- **Annealing the aux weight:** Decaying `aux_gradient_weight` toward 0 after the first 100 epochs could avoid the late-training instability while still benefiting from early-stage regularization.
- **Better gradient targets:** The kNN-based inverse-distance-weighted gradients are approximate. If the mesh connectivity is available, finite-element gradient operators would give exact targets and might avoid the noise that caused instability.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3214 because ADVISOR: Closing — both trials failed to beat baseline. Trial 1 (aux_weight=0.1): diverged ep22 from gradient explosion. Trial 2 (aux_weight=0.01): best val=5.485% ep191 (+43% worse than 3.833%), then instability event ep192. Root cause: kNN finite-difference gradient targets are inherently noisy — once primary loss flattens in late training, the noisy auxiliary signal dominates and injects harmful variance. Supervised gradient prediction is not helping when the model already learns appropriate gradient structure from the dense primary pressure loss. Good execution and clean analysis, kakashi..
